--- b/lib/boot/shims/src/platform.js.pp
+++ a/lib/boot/shims/src/platform.js.pp
@@ -1,6 +1,9 @@
 
 /* platform-specific setup */
 
+// top-level debug initialization needs this. declare it in case we aren't in the same file as out.js
+function h$ghcjszmprimZCGHCJSziPrimziJSVal_con_e() { return h$stack[h$sp]; };
+
 /*
    if browser mode is active (GHCJS_BROWSER is defined), all the runtime platform
    detection code should be removed by the preprocessor. The h$isPlatform variables
