From 34292fadaf90571dba15e84ee66eb601ab8b317f Mon Sep 17 00:00:00 2001
From: John Ericson <Ericson2314@Yahoo.com>
Date: Sat, 11 Feb 2017 21:58:56 -0500
Subject: [PATCH] Allow any arch with linux for foreign libs

---
 Cabal/Distribution/Simple/Configure.hs | 3 +--
 1 file changed, 1 insertion(+), 2 deletions(-)

diff --git a/Cabal/Distribution/Simple/Configure.hs b/Cabal/Distribution/Simple/Configure.hs
index 3c1124942..5b39204e5 100644
--- Cabal/Distribution/Simple/Configure.hs
+++ Cabal/Distribution/Simple/Configure.hs
@@ -1894,8 +1894,7 @@ checkForeignLibSupported comp platform flib = go (compilerFlavor comp)
 
     goGhcPlatform :: Platform -> Maybe String
     goGhcPlatform (Platform X86_64 OSX    ) = goGhcOsx     (foreignLibType flib)
-    goGhcPlatform (Platform I386   Linux  ) = goGhcLinux   (foreignLibType flib)
-    goGhcPlatform (Platform X86_64 Linux  ) = goGhcLinux   (foreignLibType flib)
+    goGhcPlatform (Platform _      Linux  ) = goGhcLinux   (foreignLibType flib)
     goGhcPlatform (Platform I386   Windows) = goGhcWindows (foreignLibType flib)
     goGhcPlatform (Platform X86_64 Windows) = goGhcWindows (foreignLibType flib)
     goGhcPlatform _ = unsupported [
-- 
2.14.0

