diff --git a/jsaddle.cabal b/jsaddle.cabal
index 1be651c..6a05705 100644
--- a/jsaddle.cabal
+++ b/jsaddle.cabal
@@ -24,7 +24,8 @@ library
     if impl(ghcjs -any)
         build-depends:
             ghcjs-base -any,
-            ghcjs-prim -any
+            ghcjs-prim -any,
+            ghcjs-json -any
     else
         build-depends:
             attoparsec >=0.11 && <0.14,
diff --git a/src/Language/Javascript/JSaddle/Value.hs b/src/Language/Javascript/JSaddle/Value.hs
index 3a4fff8..e00701c 100644
--- a/src/Language/Javascript/JSaddle/Value.hs
+++ b/src/Language/Javascript/JSaddle/Value.hs
@@ -110,6 +110,8 @@ import Language.Javascript.JSaddle.Marshal.String (ToJSString(..), FromJSString(
 import Language.Javascript.JSaddle.String (strToText, textToStr)
 import GHCJS.Foreign.Internal (jsTrue, jsFalse, jsNull, toJSBool, jsUndefined, isTruthy, isNull, isUndefined)
 
+import JavaScript.JSON.Types.ToJSVal ()
+
 -- $setup
 -- >>> import Language.Javascript.JSaddle.Test (testJSaddle)
 -- >>> import Language.Javascript.JSaddle.Monad (catch)
