diff --git a/compiler/hsSyn/HsExprBin_ConversionPS2SE.hs b/compiler/hsSyn/HsExprBin_ConversionPS2SE.hs
index 401f431b77c1e409766c70683114ef4f0ef85f34..7e2b6e957a66afdec89ef5bae525acfe2a834688 100644
--- a/compiler/hsSyn/HsExprBin_ConversionPS2SE.hs
+++ b/compiler/hsSyn/HsExprBin_ConversionPS2SE.hs
@@ -488,7 +488,7 @@ cvMatch f (Match a b c d) = Match a
    <$> convertName b <*> traverse (traverse cvPat) c <*> cvGRHSs f d
 cvMatch _ (XMatch a) = pure (XMatch a)
 
-cvLHsPat :: LHsPat GhcPs -> Conv (LHsPat GhcSe)
+cvLHsPat :: LPat GhcPs -> Conv (LPat GhcSe)
 cvLHsPat = traverse cvPat
 
 cvPat :: Pat GhcPs -> Conv (Pat GhcSe)
diff --git a/compiler/hsSyn/HsExprBin_ConversionSE2PS.hs b/compiler/hsSyn/HsExprBin_ConversionSE2PS.hs
index cea4c619605cde2a288840c782bba1c3076b8cac..cf39727f10224dfd74283d4a6d22fd64005e676b 100644
--- a/compiler/hsSyn/HsExprBin_ConversionSE2PS.hs
+++ b/compiler/hsSyn/HsExprBin_ConversionSE2PS.hs
@@ -489,7 +489,7 @@ cvMatch f (Match a b c d) = Match a
    <$> convertName b <*> traverse (traverse cvPat) c <*> cvGRHSs f d
 cvMatch _ (XMatch a) = pure (XMatch a)
 
-cvLHsPat :: LHsPat GhcSe -> Conv (LHsPat GhcPs)
+cvLHsPat :: LPat GhcSe -> Conv (LPat GhcPs)
 cvLHsPat = traverse cvPat
 
 cvPat :: Pat GhcSe -> Conv (Pat GhcPs)
