diff --git a/attoparsec.cabal b/attoparsec.cabal
index ecbd5db..4e57b1f 100644
--- a/attoparsec.cabal
+++ b/attoparsec.cabal
@@ -34,8 +34,8 @@ Flag developer
 
 -- We need to test and benchmark these modules,
 -- but do not want to expose them to end users
-library attoparsec-internal
-  hs-source-dirs: internal
+library
+  hs-source-dirs: internal, .
   build-depends: array,
                  base >= 4.3 && < 5,
                  bytestring <0.12,
@@ -51,7 +51,6 @@ library attoparsec-internal
   ghc-options: -O2 -Wall
   default-language: Haskell2010
 
-library
   build-depends: array,
                  base >= 4.3 && < 5,
                  bytestring <0.12,
@@ -60,8 +59,7 @@ library
                  scientific >= 0.3.1 && < 0.4,
                  transformers >= 0.2 && (< 0.4 || >= 0.4.1.0) && < 0.7,
                  text >= 1.1.1.3,
-                 ghc-prim <0.9,
-                 attoparsec-internal
+                 ghc-prim <0.9
   if impl(ghc < 7.4)
     build-depends:
       bytestring < 0.10.4.0
@@ -120,7 +118,6 @@ test-suite attoparsec-tests
   build-depends:
     array,
     attoparsec,
-    attoparsec-internal,
     base,
     bytestring,
     deepseq >= 1.1,
@@ -167,7 +164,6 @@ benchmark attoparsec-benchmarks
   build-depends:
     array,
     attoparsec,
-    attoparsec-internal,
     base == 4.*,
     bytestring >= 0.10.4.0,
     case-insensitive,
