diff --git a/Data/Hashable/Class.hs b/Data/Hashable/Class.hs
index 93470db..f641d40 100644
--- a/Data/Hashable/Class.hs
+++ b/Data/Hashable/Class.hs
@@ -1,5 +1,8 @@
 {-# LANGUAGE BangPatterns, CPP, ForeignFunctionInterface, MagicHash,
              ScopedTypeVariables, UnliftedFFITypes #-}
+#ifdef __GHCJS__
+{-# LANGUAGE JavaScriptFFI, UnboxedTuples, GHCForeignImportPrim #-}
+#endif
 
 #if __GLASGOW_HASKELL__ < 710
 {-# LANGUAGE DeriveDataTypeable #-}
@@ -75,7 +78,9 @@ import Data.Int (Int8, Int16, Int32, Int64)
 import Data.List (foldl')
 import Data.Ratio (Ratio, denominator, numerator)
 import qualified Data.Text as T
+#ifndef __GHCJS__
 import qualified Data.Text.Array as TA
+#endif
 import qualified Data.Text.Internal as T
 import qualified Data.Text.Lazy as TL
 import Data.Version (Version(..))
@@ -90,6 +95,9 @@ import GHC.Prim (ThreadId#)
 import System.IO.Unsafe (unsafeDupablePerformIO)
 import System.Mem.StableName
 import Data.Unique (Unique, hashUnique)
+#ifdef __GHCJS__
+import Data.JSString (JSString)
+#endif
 
 -- As we use qualified F.Foldable, we don't get warnings with newer base
 import qualified Data.Foldable as F
@@ -156,7 +164,11 @@ import qualified Data.ByteString.Short.Internal as BSI
 #  define MIN_VERSION_integer_gmp_1_0_0
 # endif
 
+#ifndef __GHCJS__
 import GHC.Exts (Int(..))
+#else
+import GHC.Exts (Int(..), Int#)
+#endif
 import GHC.Integer.GMP.Internals (Integer(..))
 # if defined(MIN_VERSION_integer_gmp_1_0_0)
 import GHC.Exts (sizeofByteArray#)
@@ -629,9 +641,15 @@ instance Hashable BSI.ShortByteString where
 #endif
 
 instance Hashable T.Text where
+#ifndef __GHCJS__
     hashWithSalt salt (T.Text arr off len) =
         hashByteArrayWithSalt (TA.aBA arr) (off `shiftL` 1) (len `shiftL` 1)
         salt
+#else
+    hashWithSalt salt (T.Text txt) =
+        let (# ba, len #) = js_textFromJSString txt
+        in hashByteArrayWithSalt ba (0 `shiftL` 1) (I# len `shiftL` 1) salt
+#endif
 
 instance Hashable TL.Text where
     hashWithSalt = TL.foldlChunks hashWithSalt
@@ -829,6 +847,12 @@ instance Hashable a => Hashable (Option a) where
     hashWithSalt p (Option a) = hashWithSalt p a
 #endif
 
+#ifdef __GHCJS__
+foreign import javascript unsafe
+  "h$textFromString"
+  js_textFromJSString :: JSString -> (# ByteArray#, Int# #)
+#endif
+
 -- instances for @Data.Functor.{Product,Sum,Compose}@, present
 -- in base-4.9 and onward.
 #if MIN_VERSION_base(4,9,0)
@@ -917,4 +941,3 @@ instance Show1 Hashed where
   liftShowsPrec sp _ d (Hashed a _) = showsUnaryWith sp "hashed" d a
 #endif
 
-
diff --git a/hashable.cabal b/hashable.cabal
index 74f6512..4182f38 100644
--- a/hashable.cabal
+++ b/hashable.cabal
@@ -51,6 +51,9 @@ Library
   Build-depends:     base >= 4.4 && < 4.12,
                      bytestring >= 0.9 && < 0.11,
                      deepseq >= 1.3 && < 1.5
+
+  if impl(ghcjs)
+    Build-depends:   ghcjs-base
   if impl(ghc)
     Build-depends:   ghc-prim,
                      text >= 0.11.0.5
