diff --git a/jsaddle.cabal b/jsaddle.cabal
index ce5c65f..62afdcd 100644
--- a/jsaddle.cabal
+++ b/jsaddle.cabal
@@ -32,7 +32,8 @@ library
     if impl(ghcjs -any)
         build-depends:
             ghcjs-base -any,
-            ghcjs-prim -any
+            ghcjs-prim -any,
+            ghcjs-json -any
     else
         build-depends:
             attoparsec >=0.11 && <0.14,
diff --git a/src/Language/Javascript/JSaddle/Value.hs b/src/Language/Javascript/JSaddle/Value.hs
index 6ead633..afc0241 100644
--- a/src/Language/Javascript/JSaddle/Value.hs
+++ b/src/Language/Javascript/JSaddle/Value.hs
@@ -36,6 +36,8 @@ module Language.Javascript.JSaddle.Value (
   -- * JavaScript value references
     JSVal
   , ToJSVal(..)
+  , module JavaScript.JSON.Types.FromJSVal
+  , module JavaScript.JSON.Types.ToJSVal
 
   -- * Haskell types for JavaScript values
   , JSNull(..)
@@ -110,6 +112,9 @@ import Language.Javascript.JSaddle.Marshal.String (ToJSString(..), FromJSString(
 import Language.Javascript.JSaddle.String (strToText, textToStr)
 import GHCJS.Foreign.Internal (jsTrue, jsFalse, jsNull, toJSBool, jsUndefined, isTruthy, isNull, isUndefined)
 
+import JavaScript.JSON.Types.FromJSVal ()
+import JavaScript.JSON.Types.ToJSVal ()
+
 -- $setup
 -- >>> import Language.Javascript.JSaddle.Test (testJSaddle)
 -- >>> import Language.Javascript.JSaddle.Monad (catch)
